DPE-2390 upgrade rollback test

paulomach opened the pull request:
## Issue

Additionally to upgrade happy path, we need test rollback and upgrade from `8.0/stable`

## To make it work:

**Upgrade from 8.0/stable - revision 151**
1. fix permission ownership for `/var/snap/charmed-mysql/common` from root to snap daemon after snap refresh ✅
2. add flag to bypass lack of `installed-by-charmed-mysql` file, avoid multiple snaps exceptions ✅
3. create `upgrade-peer-relation` data:
	1. each unit add `state="ready"` ✅
	2. leader unit populates `upgrade_stack` ✅
	3. leader unit populate `dependencies` structure ✅
4. leader unit runs `charm.leader-elected` hook to create missing backup user password ✅
5. leader unit remove `root@%` user ✅
6. leader unit create `backup@%` user 

**Upgrade from 8.0/edge**
1. No adaptations

paulomach commented on src/mysql_vm_helpers.py:
```diff
@@ -166,11 +170,19 @@ def install_and_configure_mysql_dependencies() -> None:
+            # fix ownership necessary for upgrades from 8/stable@r151
+            if common_path.owner() != MYSQL_SYSTEM_USER:
+                logger.debug("Updating charmed-mysql common directory ownership")
+                os.system(
+                    f"chown -R {MYSQL_SYSTEM_USER}:{MYSQL_SYSTEM_USER} {CHARMED_MYSQL_COMMON_DIRECTORY}"
+                )
```
`shutil.chown` has no recursion - either case, this may migrate to the snap instead

paulomach commented on src/mysql_vm_helpers.py:
```diff
@@ -166,11 +170,19 @@ def install_and_configure_mysql_dependencies() -> None:
                 mysqlsh_help_command = ["charmed-mysql.mysqlsh", "--help"]
                 subprocess.check_call(mysqlsh_help_command, stderr=subprocess.PIPE)
 
+            # fix ownership necessary for upgrades from 8/stable@r151
+            if common_path.owner() != MYSQL_SYSTEM_USER:
+                logger.debug("Updating charmed-mysql common directory ownership")
+                os.system(
+                    f"chown -R {MYSQL_SYSTEM_USER}:{MYSQL_SYSTEM_USER} {CHARMED_MYSQL_COMMON_DIRECTORY}"
+                )
```
see previous comment. this is probably a learning step before the final solution

paulomach replied:
that most probably this can be moved to the snap

paulomach replied:
https://github.com/canonical/charmed-mysql-snap/pull/38

paulomach commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -59,31 +86,148 @@ async def test_pre_upgrade_check(ops_test: OpsTest, mysql_charm_series: str) ->
+    # backup charm file for rollback test
+    if not isinstance(charm, Path):
+        charm = Path(charm)
+    charm.rename(f"/tmp/{charm.name}-backup")
```
on the making

paulomach replied:
ad366ca

paulomach commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -59,31 +86,148 @@ async def test_pre_upgrade_check(ops_test: OpsTest, mysql_charm_series: str) ->
+async def inject_dependency_fault(ops_test: OpsTest, application_name: str):
+    """Inject a dependency fault into the mysql charm."""
+    # Open dependency.json and load current charm version
+    if os.path.exists("src/dependency.json-orig"):
+        # restore original file for multiple test runs
+        copy("src/dependency.json-orig", "src/dependency.json")
+    else:
+        # backup original file
+        copy("src/dependency.json", "src/dependency.json-orig")
```
new solution won't need this anymore

paulomach replied:
ad366ca

paulomach commented on src/mysql_vm_helpers.py:
```diff
@@ -152,7 +152,11 @@ def install_and_configure_mysql_dependencies() -> None:
             CHARMED_MYSQL_COMMON_DIRECTORY, "installed_by_mysql_server_charm"
         )
-        if charmed_mysql.present and not installed_by_mysql_server_file.exists():
+        if (
+            charmed_mysql.present
+            and not installed_by_mysql_server_file.exists()
+            and not ignore_multiple
+        ):
```
The problem is that we started creating the file on rev169 only. Suggestions?

paulomach replied:
The issue here is that even if the running revision is known (which I'm not entirely sure it's possible from the charm), when installing from local charm file the revision number is meaningless. 
One alternative would be add the need to create a file manually into the documentation.

paulomach commented on lib/charms/mysql/v0/mysql.py:
```diff
@@ -476,6 +477,14 @@ def unit_peer_data(self) -> Dict:
+        if not self.peers:
+            return set()
+
+        return set([self.unit] + list(self.peers.units))
```
ad366ca

paulomach commented on lib/charms/mysql/v0/mysql.py:
```diff
@@ -476,6 +477,14 @@ def unit_peer_data(self) -> Dict:
         return self.peers.data[self.unit]
 
+    @property
+    def app_units(self) -> Set[Unit]:
```
ad366ca

paulomach commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -1,44 +1,71 @@
+@pytest.mark.abort_on_fail
+async def test_deploy_latest(ops_test: OpsTest) -> None:
     """Simple test to ensure that the mysql and application charms get deployed."""
-    await high_availability_test_setup(ops_test, mysql_charm_series)
```
moved back
ad366ca

paulomach commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -59,31 +86,148 @@ async def test_pre_upgrade_check(ops_test: OpsTest, mysql_charm_series: str) ->
+
+    # backup charm file for rollback test
+    if not isinstance(charm, Path):
+        charm = Path(charm)
```
ad366ca

taurus-forever commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -50,40 +80,152 @@ async def test_pre_upgrade_check(ops_test: OpsTest, mysql_charm_series: str) ->
+        num_units=3,
+        channel="8.0/stable",
+        series=mysql_charm_series,
+        # config={"profile": "testing"}, # config not available in 8.0/stable@r151
```
hint: add constraint for RAM to avoid slow GH runner due to lack of RAM.

taurus-forever reviewed (approved):
Incredible work done!

paulomach commented on pyproject.toml:
```diff
@@ -57,7 +57,7 @@ coverage = {extras = ["toml"], version = "^7.2.7"}
 pytest-github-secrets = {git = "https://github.com/canonical/data-platform-workflows", tag = "v4.2.3", subdirectory = "python/pytest_plugins/github_secrets"}
 pytest-operator = "^0.28.0"
-pytest-operator-cache = {git = "https://github.com/canonical/data-platform-workflows", tag = "v4.2.3", subdirectory = "python/pytest_plugins/pytest_operator_cache"}
+pytest-operator-cache = {git = "https://github.com/canonical/data-platform-workflows", tag = "v5.0.0", subdirectory = "python/pytest_plugins/pytest_operator_cache"}
```
b1ebaac

paulomach replied:
076481ff

paulomach commented on src/mysql_vm_helpers.py:
```diff
@@ -155,8 +158,10 @@ def install_and_configure_mysql_dependencies(ignore_multiple: bool = False) -> N
+            and not upgrade_refresh
         ):
+            # TODO: the `installed_by_*` file was introduced on revision 169 of the charm
+            # and should be removed after a few releases
```
b1ebaac

paulomach commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -29,9 +31,9 @@
 
 
-@pytest.mark.group(1)
+@pytest.mark.group(2)
```
leftover from testing: b1ebaac

paulomach commented on src/upgrade.py:
```diff
@@ -311,6 +315,15 @@ def _prepare_upgrade_from_legacy(self) -> None:
         )
 
+    @staticmethod
+    def _ensure_for_installed_by_file() -> None:
+        """Ensure snap mark file to allow refresh."""
+        installed_by_mysql_server_file = pathlib.Path(
+            CHARMED_MYSQL_COMMON_DIRECTORY, "installed_by_mysql_server_charm"
+        )
```
Not as designed for the dependency check in the upgrade lib. We do not record the revision as in the revision generated by the store, so there's not a clear mapping to revision:feature. Instead we have an arbitrary versioning semver to infer compatibility. 
It's a discussion that will have further evolution, IMO.

paulomach commented on src/upgrade.py:
```diff
@@ -128,6 +140,31 @@ def _online_instances(status_dict: dict) -> int:
             )
 
+    def _on_upgrade_charm_check_legacy(self, event) -> None:
+        if not self.peer_relation or len(self.app_units) < len(self.charm.app_units):
```
The relation is created in the refresh, so when any hook from the new charm runs, the relation is present, but it may not yet have all units joined.

paulomach commented:
> Changes since last review look good
> 
> IMO these should be addressed before merge, but feel free to ignore if others disagree:
> 
>     * remove use of global: [DPE-2390 upgrade rollback test #301 (comment)](https://github.com/canonical/mysql-operator/pull/301#discussion_r1304815757)
> 
>     * if upgrading from old version, create `installed_by_mysql_server_charm` instead of always creating if it not exists on upgrade [DPE-2390 upgrade rollback test #301 (comment)](https://github.com/canonical/mysql-operator/pull/301#discussion_r1307770136)
> 
>     * [DPE-2390 upgrade rollback test #301 (comment)](https://github.com/canonical/mysql-operator/pull/301#discussion_r1304547007)

Added to [dpe-2409](https://warthogs.atlassian.net/browse/DPE-2409)

shayancanonical commented on tests/integration/high_availability/high_availability_helpers.py:
```diff
@@ -106,7 +106,8 @@ async def ensure_n_online_mysql_members(
     except RetryError:
-        return False
+        pass
+    return False
```
why this change? it is equivalent of `return False` from the except block

paulomach replied:
just to make the linter happy

shayancanonical commented on tests/integration/high_availability/test_upgrade.py:
```diff
@@ -50,40 +80,112 @@ async def test_pre_upgrade_check(ops_test: OpsTest, mysql_charm_series: str) ->
+    loaded_dependency_dict["charm"]["version"] = f"{int(current_charm_version)+1}"
+
+    # Overwrite dependency.json with incompatible version
+    with zipfile.ZipFile(charm_file, mode="a") as charm_zip:
```
why use zipfile?

paulomach replied:
the `.charm` is a zipfile - this is very direct way to overwrite a file within a packed `.charm`

shayancanonical reviewed (commented):
Looks great, but left 2 minor questions
